honeycombio/data_source_query.go

```diff
@@ -164,31 +186,9 @@ func dataSourceHoneycombioQueryRead(ctx context.Context, d *schema.ResourceData,
 		}
 	}
 
-	filterSchemas := d.Get("filter").(*schema.Set).List()
-	filters := make([]honeycombio.FilterSpec, len(filterSchemas))
-
-	for i, f := range filterSchemas {
-		fMap := f.(map[string]interface{})
-
-		filters[i] = honeycombio.FilterSpec{
-			Column: fMap["column"].(string),
-			Op:     honeycombio.FilterOp(fMap["op"].(string)),
-			Value:  fMap["value"].(string),
-		}
-
-		// Ensure we don't send filter.Value if op is "exists" or
-		// "does-not-exist". The Honeycomb API will refuse this.
-		//
-		// TODO ideally this check is part of the schema (as a ValidateFunc),
-		//      but this is not yet supported by the SDK.
-		//      https://github.com/hashicorp/terraform-plugin-sdk/issues/155#issuecomment-489699737
-		filter := filters[i]
-		if filter.Op == honeycombio.FilterOpExists || filter.Op == honeycombio.FilterOpDoesNotExist {
-			if filter.Value != "" {
-				return diag.Errorf("Filter operation %s must not contain a value", filter.Op)
-			}
-			filters[i].Value = nil
-		}
+	filters, err := extractFilters(d)
+	if err != nil {
+		return diag.FromErr(err)
```

Comment:
this was getting a little unwieldy, felt like it was time to pull it into a function.

Not sure if it makes sense to do it like this or return diags all the way up

Reply:
Agreed, good call :+1: